src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/BinaryEncoders/CommandMessageBinaryEncoder.cs

```diff
@@ -288,25 +294,157 @@ private void WriteType1Section(BsonBinaryWriter writer, Type1CommandMessageSecti
             var payloadStartPosition = stream.Position;
             stream.WriteInt32(0); // size
             stream.WriteCString(section.Identifier);
+
             var batch = section.Documents;
-            var processedCount = batch.Count;
-            for (var i = 0; i < batch.Count; i++)
+            var maxDocumentSize = section.MaxDocumentSize ?? writer.Settings.MaxDocumentSize;
+            writer.PushSettings(s => ((BsonBinaryWriterSettings)s).MaxDocumentSize = maxDocumentSize);
+            writer.PushElementNameValidator(section.ElementNameValidator);
+            try
             {
-                var documentStartPosition = stream.Position;
-                var document = batch.Items[batch.Offset + i];
-                serializer.Serialize(context, document);
-
-                var messageSize = stream.Position - messageStartPosition;
-                if (messageSize > maxMessageSize)
+                var maxBatchCount = Math.Min(batch.Count, section.MaxBatchCount ?? int.MaxValue);
+                var processedCount = maxBatchCount;
+                for (var i = 0; i < maxBatchCount; i++)
                 {
-                    stream.Position = documentStartPosition;
-                    stream.SetLength(documentStartPosition);
-                    processedCount = i;
-                    break;
+                    var documentStartPosition = stream.Position;
+                    var document = batch.Items[batch.Offset + i];
+                    serializer.Serialize(context, document);
+
+                    var messageSize = stream.Position - messageStartPosition;
+                    if (messageSize > maxMessageSize && batch.CanBeSplit)
+                    {
+                        stream.Position = documentStartPosition;
+                        stream.SetLength(documentStartPosition);
+                        processedCount = i;
+                        break;
+                    }
                 }
+                batch.SetProcessedCount(processedCount);
+            }
+            finally
+            {
+                writer.PopElementNameValidator();
+                writer.PopSettings();
             }
-            batch.SetProcessedCount(processedCount);
             stream.BackpatchSize(payloadStartPosition);
         }
+
+        // nested types
+        private class PostProcessor : IMessageEncoderPostProcessor
```

Comment:
This is what the PostProcessorAction can call to change things *after* the message is written.

So far, the only possible change that can be made is to change the write concern from { w : 0 } to { w : 1 }.